daemon/cmd/app.go

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"go.uber.org/fx"
+
+	"github.com/cilium/cilium/api/v1/server"
+)
+
+const (
+	startTimeout = 5 * time.Minute
+	stopTimeout  = 5 * time.Minute
+)
+
+// tempDaemonModule wraps the existing daemon initialization
+// and provides access to *Daemon. This module is meant
+// to be temporary for the purpose of incrementally
+// refactoring the initialization by lifting modules
+// out from the 'Daemon' structure.
+//
+// Note that runDaemon is used as constructor, but it
+// is effectively starting the application.
+var tempDaemonModule = fx.Module(
+	"temp-daemon",
+
+	fx.Provide(runDaemon),
+	fx.Invoke(func(*Daemon) {}),
+)
+
+func runApp() {
+	// Create a top-level context for the daemon. This is cancelled by the signal handler
+	// in cleanup.go and will trigger stop.
+	ctx, cancel := context.WithCancel(server.ServerCtx)
+
+	app := fx.New(
+		fx.WithLogger(newAppLogger),
+		fx.Supply(
+			fx.Annotate(ctx, fx.As(new(context.Context))),
+			cancel,
+		),
+		tempDaemonModule,
+	)
+
+	if app.Err() != nil {
+		log.WithError(app.Err()).Fatal("Failed to initialize daemon")
+	}
+
+	// Normally one would use app.Run() which takes care of starting and stopping
+	// and handles signals to trigger to stop. Since we're hooking this into
+	// existing code that already handles the signals, we're opting to manually
+	// start, then wait for the context to be cancelled by existing signal handler
+	// and then stop.
+	startCtx, startCtxCancel := context.WithTimeout(context.Background(), startTimeout)
+	defer startCtxCancel()
+
+	if err := app.Start(startCtx); err != nil {
+		if ctx.Err() != nil {
+			// Context was cancelled during startup, e.g. due to interrupt. Exit
+			// normally in this case.
+			os.Exit(0)
+		}
+		os.Exit(1)
+	}
+
+	// Wait until the daemon context is cancelled. This is done by the
+	// cleaner on receipt of an interrupt.
+	<-ctx.Done()
```

Comment:
Ah, now I see why the current implementation passes the `cancelFunc` down. I guess this works for now, though a bit odd. Maybe we can explore using [App.Done()](https://pkg.go.dev/go.uber.org/fx#App.Done) and `Shutdowner`s to do this more in line with the way that fx expects. Minor nits though, it's not super clear to me whether there's a motivation to look into this specifically now as part of this PR or consider handling it in future. May partly depend on where we think the code is eventually going to get to.

Reply:
Yep, it's a bit unsatisfying how this tries to stick with the existing way of stopping things. My take on this currently is to not yet touch these parts, but perhaps as one of the first follow ups refactor the signal handling and cleanup to work the fx way?

Reply:
Second commit now implements the cleanup properly.